rust/private/clippy.bzl

```diff
@@ -134,8 +139,43 @@ def _clippy_aspect_impl(target, ctx):
         mnemonic = "Clippy",
     )
 
+    # A second run of clippy, this time to capture the JSON output.
+    clippy_json = ctx.actions.declare_file(ctx.label.name + ".clippy.json")
```

Comment:
Is it possible to have just one action that generates this json file? I'm assuming the reason this ended up as two was to make sure the clippy aspect continued to show logs but also generate the json file? If it can't be merged into a single action, I'd say perhaps we shouldn't be using a provider here and instead allow users to request that this json file be created by selecting the output group they're interested in? What do you think?

Reply:
So, what this change gives us, which I think we want, is two ways to run clippy. The existing way prints human-readable output to STDOUT. In that mode, all findings are treated as errors, as a hack to force the findings to always be printed when the build is re-run. The only output from Bazel's perspective is a touch file.

But when generating JSON for consumption by other tools, storing the output is what we want for the sake of efficiency. I didn't see a better way to unify the two ways of invoking clippy, other than copy-and-paste, which I don't like.

I think, though, that the ClippyOutputInfo does let the user select what they are interested in.

(The particular motivating use case for us is to have another aspect consume the ClippyOutputInfo and surface that in our code review tool)

Reply:
Cool! the thing I care most about is not running the actions unnecessarily. Can you explain why an [OutputGroupInfo](https://docs.bazel.build/versions/main/skylark/lib/OutputGroupInfo.html) might not be usable in a downstream aspect? (Sorry if I'm blanking on something obvious 😅)

Reply:
Hmm...maybe I can use an OutputGroupInfo. I'll give it a try. I'm definitely not a Starlark expert. The aspect-on-aspect stuff is apparently still not well-supported in Bazel, so it hasn't always been clear whether I just don't know what I'm doing, or if there's a limitation in Bazel. Fun times.

Reply:
Fun times indeed... I'm happy to help work through issues however I can though!

Maybe sharing an example of the aspect you have will help provide better context and inspire some ideas. But understandable if it's not sharable.

Reply:
There's nothing particularly sensitive about it. Essentially all it does is consume the clippy JSON output and convert it to a protobuf message:

```
load("//rust/private:clippy.bzl", "ClippyOutputInfo")

def _clippy_aspect_impl(target, ctx):
    output = ctx.actions.declare_file(ctx.label.name + ".clippy.binarypb")
    ctx.actions.run(
        executable = ctx.executable._converter,
        outputs = [output],
        inputs = target[ClippyOutputInfo].json,
        arguments = ["--input_file", target[ClippyOutputInfo].json.to_list()[0].path, "--output_file", output.path],
    )
    return [
        OutputGroupInfo(<magical output group name> = depset([output])),
    ]

clippy_aspect = aspect(
    attrs = {
        "_converter": attr.label(
            doc = "Binary to convert clippy JSON output to protobuf",
            default = Label("//blah/blah:converter"),
            executable = True,
            cfg = "exec",
        ),
    },
    implementation = _clippy_aspect_impl,
    required_aspect_providers = [ClippyOutputInfo],
)
```

Reply:
Ah yeah, that seems totally reasonable and correct. I don't otherwise know how to require that particular output at the aspect definition level. Thanks for sharing!